src/metric-engine/src/engine/read.rs

```diff
@@ -0,0 +1,251 @@
+use crate::utils;
+
+impl MetricEngineInner {
+    #[tracing::instrument(skip_all)]
+    pub async fn read_region(
+        &self,
+        region_id: RegionId,
+        request: ScanRequest,
+    ) -> Result<SendableRecordBatchStream> {
+        let is_reading_physical_region = self
+            .state
+            .read()
+            .await
+            .physical_regions()
+            .contains_key(&region_id);
+
+        if is_reading_physical_region {
+            info!(
+                "Metric region received read request {request:?} on physical region {region_id:?}"
+            );
+            self.read_physical_region(region_id, request).await
+        } else {
+            self.read_logical_region(region_id, request).await
+        }
+    }
+
+    /// Proxy the read request to underlying physical region (mito engine).
+    async fn read_physical_region(
+        &self,
+        region_id: RegionId,
+        request: ScanRequest,
+    ) -> Result<SendableRecordBatchStream> {
+        self.mito
+            .handle_query(region_id, request)
+            .await
+            .context(MitoReadOperationSnafu)
+    }
+
+    async fn read_logical_region(
+        &self,
+        logical_region_id: RegionId,
+        request: ScanRequest,
+    ) -> Result<SendableRecordBatchStream> {
+        let physical_region_id = {
+            let state = &self.state.read().await;
+            state
+                .get_physical_region_id(logical_region_id)
+                .with_context(|| {
+                    error!("Trying to alter an nonexistent region {logical_region_id}");
+                    LogicalRegionNotFoundSnafu {
+                        region_id: logical_region_id,
+                    }
+                })?
+        };
+        let data_region_id = utils::to_data_region_id(physical_region_id);
+        let request = self
+            .transform_request(physical_region_id, logical_region_id, request)
+            .await?;
+        self.mito
+            .handle_query(data_region_id, request)
+            .await
+            .context(MitoReadOperationSnafu)
+    }
+
+    /// Transform the [ScanRequest] from logical region to physical data region.
+    async fn transform_request(
+        &self,
+        physical_region_id: RegionId,
+        logical_region_id: RegionId,
+        mut request: ScanRequest,
+    ) -> Result<ScanRequest> {
+        // transform projection
+        let physical_projection = if let Some(projection) = &request.projection {
+            self.transform_projection(physical_region_id, logical_region_id, projection)
+                .await?
+        } else {
+            self.default_projection(physical_region_id, logical_region_id)
+                .await?
+        };
+        request.projection = Some(physical_projection);
+
+        // add table filter
+        request
+            .filters
+            .push(self.table_id_filter(logical_region_id));
```

Comment:
Now mito doesn't filter rows exactly so it is possible to return rows from other logical tables. We need to support the exact filter later (in mito or metrics engine).